pylds/base.py

```diff
@@ -6,32 +6,122 @@
 Publisher, Number(Volume No), pp.142-161.
 """
 
+import sys
 import numpy as np
-import matplotlib.pyplot as plt
+import numpy.ma as ma
+from operator import itemgetter
 from scipy.integrate import solve_ivp
+from functools import reduce
+
+def energy_conservation_condition(phase_space_axes, H0, potential_energy, momentum_sign):
+    N_dof = len(phase_space_axes)
+    phase_space_positions = np.transpose(phase_space_axes[:int(N_dof/2)])
+    phase_space_momenta = np.transpose(phase_space_axes[int(N_dof/2):])
+    V = potential_energy(phase_space_positions)
+    points_dof_remaining = momentum_sign * \
+        np.sqrt(2*(H0 - V) - (phase_space_momenta**2).sum(axis=1))
+    return points_dof_remaining
 
 def generate_points(grid_parameters):
     """
     Returns a 1D array of all points from a on a uniform grid with dimensions and size defined by list of input parameters.
     An additional dimension initiallised with zeros is added for the calculation of Lagrangian Descriptors.
-    
+
     Parameters
     ----------
-    grid_parameters : list of 3-tuples of floats
+    slice_parameters : list of 3-tuples of floats
         input parameters of limits and size of mesh per axis
-        
+
     Returns
     -------
     mesh : 1d numpy array
         flattened array of initial conditions
     """
-    x_min, x_max, Nx = grid_parameters[0]
-    y_min, y_max, Ny = grid_parameters[1]
-    points_x = np.linspace(x_min, x_max, Nx)
-    points_y = np.linspace(y_min, y_max, Ny)    
-    Y, X = np.meshgrid(points_y, points_x)  # Grid in phase-space
-    mesh = np.transpose([X.flatten(), Y.flatten(), np.zeros(Nx*Ny)]) # 2D grid + a zero column for LDs
-    return mesh.flatten()
+    if type(grid_parameters) == dict:
+        # Unpack extra grid parameters
+        slice_parameters = grid_parameters['slice_parameters']
+        dof_slice = grid_parameters['dof_slice']
+        dof_fixed, dof_fixed_values = grid_parameters['dof_fixed']
+        momentum_sign = grid_parameters['momentum_sign']
+        potential_energy = grid_parameters['potential_energy']
+        H0 = grid_parameters['energy_level']
+
+        N_dof = len(dof_slice)  # Total number of DoF
+        # Check N DoF is even.
+        if N_dof % 2 != 0:
```

Comment:
N_dof should be renamed. It is the number of dimensions, not degrees of freedom. It seems that other variables may need renaming for the same reason

Reply:
Done!

Reply:
There are still some variables named using `dof ` in `LD_Python_Examples.ipynb`

Reply:
and in base.py, e.g. `dof_slice` mention in line 49